ecosystem/indexer-grpc/indexer-grpc-utils/src/in_memory_cache.rs

```diff
@@ -0,0 +1,439 @@
+            if current_latest_version == in_cache_latest_version {
+                tokio::select! {
+                    _ = cancellation_token.cancelled() => {
+                        tracing::info!("In-memory cache update task is cancelled.");
+                        return;
+                    },
+                    _ = tokio::time::sleep(std::time::Duration::from_millis(IN_MEMORY_CACHE_LOOKUP_RETRY_INTERVAL_MS)) => {
+                        continue;
+                    },
+                }
+            }
+            let redis_waiting_duration = current_time.elapsed().as_secs_f64();
+            let start_time = std::time::Instant::now();
+            let end_version = std::cmp::min(
+                current_latest_version,
+                in_cache_latest_version + 10 * MAX_REDIS_FETCH_BATCH_SIZE as u64,
+            );
+            let versions_to_fetch = (in_cache_latest_version..end_version).collect();
+            let transactions = batch_get_transactions(&mut conn, versions_to_fetch, storage_format)
+                .await
+                .unwrap();
+            // Ensure that transactions are ordered by version.
+            let cache_processing_start_time = std::time::Instant::now();
+            let mut newly_added_bytes = 0;
+            for (ind, transaction) in transactions.iter().enumerate() {
+                if transaction.version != in_cache_latest_version + ind as u64 {
+                    panic!("Transactions are not ordered by version");
+                }
+                newly_added_bytes += transaction.encoded_len() as u64;
+            }
+
+            for transaction in transactions {
+                cache.insert(transaction.version, Arc::new(transaction));
+            }
+            let processing_duration = start_time.elapsed().as_secs_f64();
+            tracing::info!(
+                redis_latest_version = current_latest_version,
+                in_memory_latest_version = in_cache_latest_version,
+                new_in_memory_latest_version = end_version,
+                processing_duration,
+                cache_processing_duration = cache_processing_start_time.elapsed().as_secs_f64(),
+                redis_waiting_duration,
+                "In-memory cache is updated"
+            );
+            let start_time = std::time::Instant::now();
+            let mut current_cache_metadata = { *cache_metadata.read().await };
+            current_cache_metadata.latest_version = end_version;
+            current_cache_metadata.total_size_in_bytes += newly_added_bytes;
+            // Get the data available.
+            *cache_metadata.write().await = current_cache_metadata;
+            // Clean up.
+            while current_cache_metadata.total_size_in_bytes > MAX_IN_MEMORY_CACHE_CAPACITY_IN_BYTES
+            {
+                let key_to_remove = current_cache_metadata.first_version;
+                let (_k, v) = cache
```

Comment:
because of the `v` is behind the `arc`, my assumption is it'll be dropped after the references are dropped? no in the map and not in the process of `get_transactions`, right?

Reply:
It's either here or in the get_transactions, in both cases you don't want the drop to block the foreground thread right?

Reply:
oh i see what do you mean; yea I agree that having a dedicated thread can help both the map GC and the `Arc<T>` drop.

Reply:
done.